cpp/src/ts/batched_kalman.cu

```diff
@@ -0,0 +1,820 @@
+        // MA is plus
+        // param < 1+sum(param)
+        d_new_params[n - i - 1 + ib * n] =
+          fmin((1 + sum) * eps, d_new_params[n - i - 1 + ib * n]);
+        // param > -(1+sum(param))
+        d_new_params[n - i - 1 + ib * n] =
+          fmax(-(1 + sum) * eps, d_new_params[n - i - 1 + ib * n]);
+      }
+    }
+  });
+}
+
+void unpack(const double* d_params, double* d_mu, double* d_ar, double* d_ma,
+            int batchSize, int p, int d, int q) {
+  int N = (p + d + q);
+  auto counting = thrust::make_counting_iterator(0);
+  thrust::for_each(counting, counting + batchSize, [=] __device__(int bid) {
+    if (d > 0) d_mu[bid] = d_params[bid * N];
+    for (int ip = 0; ip < p; ip++) {
+      d_ar[p * bid + ip] = d_params[bid * N + d + ip];
+    }
+    for (int iq = 0; iq < q; iq++) {
+      d_ma[q * bid + iq] = d_params[bid * N + d + p + iq];
+    }
+  });
+}
+
+void pack(int batchSize, int p, int d, int q, const double* d_mu,
+          const double* d_ar, const double* d_ma, double* d_params) {
+  int N = (p + d + q);
+  auto counting = thrust::make_counting_iterator(0);
+  thrust::for_each(counting, counting + batchSize, [=] __device__(int bid) {
+    if (d > 0) d_params[bid * N] = d_mu[bid];
+    for (int ip = 0; ip < p; ip++) {
+      d_params[bid * N + d + ip] = d_ar[p * bid + ip];
+    }
+    for (int iq = 0; iq < q; iq++) {
+      d_params[bid * N + d + p + iq] = d_ma[q * bid + iq];
+    }
+  });
+}
+
+void batched_jones_transform(cumlHandle& handle, int p, int d, int q,
+                             int batchSize, bool isInv, const double* h_params,
+                             double* h_Tparams) {
+  int N = p + d + q;
+  auto alloc = handle.getDeviceAllocator();
+  auto stream = handle.getStream();
+  double* d_params =
+    (double*)alloc->allocate(N * batchSize * sizeof(double), stream);
+  double* d_Tparams =
+    (double*)alloc->allocate(N * batchSize * sizeof(double), stream);
+  double* d_mu =
+    (double*)alloc->allocate(d * batchSize * sizeof(double), stream);
+  double* d_ar =
+    (double*)alloc->allocate(p * batchSize * sizeof(double), stream);
+  double* d_ma =
+    (double*)alloc->allocate(q * batchSize * sizeof(double), stream);
+
+  MLCommon::updateDevice(d_params, h_params, N * batchSize, stream);
+
+  unpack(d_params, d_mu, d_ar, d_ma, batchSize, p, d, q);
+
+  double* d_Tar;
+  double* d_Tma;
+  batched_jones_transform(handle, p, q, batchSize, isInv, d_ar, d_ma, d_Tar,
+                          d_Tma);
+
+  pack(batchSize, p, d, q, d_mu, d_Tar, d_Tma, d_Tparams);
+
+  MLCommon::updateHost(h_Tparams, d_Tparams, N * batchSize, stream);
+
+  alloc->deallocate(d_params, N * batchSize * sizeof(double), stream);
+  alloc->deallocate(d_Tparams, N * batchSize * sizeof(double), stream);
+  alloc->deallocate(d_mu, d * batchSize * sizeof(double), stream);
+  alloc->deallocate(d_ar, p * batchSize * sizeof(double), stream);
+  alloc->deallocate(d_ma, q * batchSize * sizeof(double), stream);
+}
```

Comment:
`d_Tar` and `d_Tma` are being allocated in `batched_jones_transform`, but not getting freed here?

Reply:
I've moved the allocations and deallocations to the same places, and changed the function API.